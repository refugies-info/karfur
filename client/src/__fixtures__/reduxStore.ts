import { RootState } from "services/rootReducer";
import { SearchResultsState } from "services/SearchResults/searchResults.reducer";
import { UserFavoritesState } from "services/UserFavoritesInLocale/UserFavoritesInLocale.reducer";
import { activeThemesMock } from "./activeThemes";

const initialMockUserState = {
  user: null,
  admin: false,
  traducteur: false,
  expertTrad: false,
  contributeur: false,
  membreStruct: false,
  userId: "",
  userFetched: false,
  rolesInStructure: [],
};
const initialMockLangueState = {
  langues: [],
  languei18nCode: "fr",
  showLanguageModal: false,
  showLangModal: false,
};

const initialMockTranslationState = {
  translation: {
    initialText: {},
    translatedText: {},
  },
  translations: [],
};

const initialUserFavoritesState: UserFavoritesState = {
  favorites: [],
  showFavoriteModal: false
}

const initialMockSearchReults: SearchResultsState = {
  results: {
    dispositifs: [],
    demarches: [],
    dispositifsSecondaryTheme: []
  },
  query: {
    search: "",
    departments: [],
    themes: [],
    needs: [],
    age: [],
    frenchLevel: [],
    language: [],
    sort: "date",
    type: "all",
  }
}

export const initialMockStore = {
  user: initialMockUserState,
  langue: initialMockLangueState,
  activeDispositifs: [],
  tts: { ttsActive: false, showAudioSpinner: false },
  userStructure: null,
  selectedDispositif: null,
  loadingStatus: {},
  translation: initialMockTranslationState,
  activeStructures: [],
  selectedStructure: null,
  allDispositifs: [],
  allStructures: [],
  users: [],
  userFavorites: initialUserFavoritesState,
  userContributions: [],
  needs: [],
  themes: {
    activeThemes: activeThemesMock,
    inactiveThemes: []
  },
  searchResults: initialMockSearchReults
};
