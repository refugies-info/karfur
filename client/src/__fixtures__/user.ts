import { GetUserInfoResponse, UserStatus } from "api-types";
import moment from "moment";

const date = moment("2019-05-01 12:00:00");
export const testUser: GetUserInfoResponse = {
  username: "test",
  _id: "testObjectId",
  structures: [],
  contributions: [],
  phone: "",
  roles: [],
  email: "",
  selectedLanguages: [],
  status: UserStatus.ACTIVE,
  traductionsFaites: []
};

export const testUserWithRoles: GetUserInfoResponse = {
  ...testUser,
  structures: ["testObjectId"],
  roles: [
    { nom: "Admin", _id: "testObjectId", nomPublic: "Admin" },
    { nom: "Trad", _id: "testObjectId", nomPublic: "Trad" },
    {
      nom: "ExpertTrad",
      _id: "testObjectId",
      nomPublic: "ExpertTrad"
    },
    {
      nom: "Contrib",
      _id: "testObjectId",
      nomPublic: "Contrib"
    }
  ]
};
