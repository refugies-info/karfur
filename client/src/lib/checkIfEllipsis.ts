export const checkIfEllipsis = (element: HTMLElement) => element.offsetWidth < element.scrollWidth;
