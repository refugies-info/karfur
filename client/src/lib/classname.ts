export const cls = (...classNames: (string | boolean)[]) => classNames.filter(className => !!className).join(" ");
