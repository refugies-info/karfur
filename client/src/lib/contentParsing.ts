export const getCalloutTranslationKey = (level: "info" | "important") => {
  switch (level) {
    case "info":
      return "callout_info"
    case "important":
      return "callout_important"
  }
}

interface ToParse {
  nodeAttr: string;
  translation: string;
}

export const translationParsing = (originalHTML: string, toParse: ToParse[]) => {
  let parsedHTML = originalHTML;
  for (const parsing of toParse) {
    parsedHTML = parsedHTML.replace(parsing.nodeAttr, `${parsing.nodeAttr} data-title='${parsing.translation}'`)
  }
  return parsedHTML
}
