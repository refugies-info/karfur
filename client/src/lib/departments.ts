import { departments } from "data/departments"

export const getDbDepartment = (depName: string) => {
  const index = departments.map(dep => dep.split(" - ")[1]).indexOf(depName);
  return index > -1 ? departments[index] : depName;
}

/**
 * Get department name or postal code from GMap autocomplete
 * @returns a department number or name formatted
 */
export const formatDepartment = (depName: string) => {
  let depDbFormat = depName;

  // postal code
  if (depName.match(/\d{5}/g)) return depName.slice(0, 2);

  // department
  if (!depDbFormat.includes(" - ")) {
    depDbFormat = getDbDepartment(depName);
  }

  const dep = depDbFormat.split(" - ");
  if (!dep[1]) return dep[0];
  return `${dep[1]} (${dep[0]})`;
}

export const getDepartmentFromNumber = (number: string): string => {
  return departments.find(d => d.startsWith(number)) || "";
}
