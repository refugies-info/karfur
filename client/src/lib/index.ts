export const jsUcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toUpperCase() + string.slice(1, string.length);

export const jsLcfirst = (string: string) =>
  string &&
  string.length > 1 &&
  string.charAt(0).toLowerCase() + string.slice(1, string.length);

export const limitNbCaracters = (string: string, nbCaractersMax: number) =>
  string.substring(0, Math.min(string.length, nbCaractersMax)) +
  (string.length > nbCaractersMax ? "..." : "");

export const removeAccents = (str = "") => {
  var accents =
    "ÀÁÂÃÄÅàáâãäåÒÓÔÕÕÖØòóôõöøÈÉÊËèéêëðÇçÐÌÍÎÏìíîïÙÚÛÜùúûüÑñŠšŸÿýŽž";
  var accentsOut =
    "AAAAAAaaaaaaOOOOOOOooooooEEEEeeeeeCcDIIIIiiiiUUUUuuuuNnSsYyyZz";
  const chars = str.split("");
  let i, x;
  for (i = 0; i < chars.length; i++) {
    if ((x = accents.indexOf(chars[i])) !== -1) {
      chars[i] = accentsOut[x];
    }
  }
  return chars.join("");
};

// rules for password check :
// nbNumUpper = nbNumeric + nbUpper
// score 0 (cannot set password): length < 5
// score 0.5 : 7>length>=5 + nbNumUpper = 1 or 2 or0
// score 1.5 : 7>length >= 5 + nbNumUpper>=3 OR length>=7 + nbNumUpper<3
// score 2.5 : length >= 7 + nbNumUpper>=3
// score 3.5 : length >= 7 + nbNumUpper>=3 + at least 1 special caractere

// make sure to use same function in backend
export const computePasswordStrengthScore = (
  password: string
): { score: number } => {
  if (password.length < 5) {
    return { score: 0 };
  }

  const numerics = password.match(/\d+/g)?.join("");
  const nbNumeric = numerics ? numerics.length : 0;

  const upperCase = password.match(/[A-Z]/g)?.join("");
  const nbUpperCase = upperCase ? upperCase.length : 0;

  if (password.length < 7) {
    if (nbUpperCase + nbNumeric < 3) {
      return { score: 1 };
    }
    return { score: 2 };
  }

  if (nbNumeric + nbUpperCase < 3) {
    return { score: 2 };
  }
  const specialCar = password.match(/[!@#$&*]/g)?.join("");
  const nbSpecialCar = specialCar ? specialCar.length : 0;
  if (nbSpecialCar > 0) {
    return { score: 4 };
  }

  return { score: 3 };
};
