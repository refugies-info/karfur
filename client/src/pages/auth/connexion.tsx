import { ReactElement, useMemo, useState } from "react";
import { useAsyncFn } from "react-use";
import { useRouter } from "next/router";
import { Button } from "@codegouvfr/react-dsfr/Button";
import { Tag } from "@codegouvfr/react-dsfr/Tag";
import { PasswordInput } from "@codegouvfr/react-dsfr/blocks/PasswordInput";
import { getPath } from "routes";
import { useAuthRedirect, useLogin } from "hooks";
import API from "utils/API";
import { defaultStaticProps } from "lib/getDefaultStaticProps";
import { cls } from "lib/classname";
import { Event } from "lib/tracking";
import SEO from "components/Seo";
import Layout from "components/Pages/auth/Layout";
import Loader from "components/Pages/auth/Loader";
import FRLink from "components/UI/FRLink";
import styles from "scss/components/auth.module.scss";

const AuthLogin = () => {
  useAuthRedirect(true);
  const router = useRouter();
  const { logUser, handleError } = useLogin();
  const email: string = useMemo(() => router.query.email as string, [router.query]);
  const has2FA = useMemo(() => router.query["2fa"], [router.query]);
  const [error, setError] = useState("");
  const [isLoading, setIsLoading] = useState(false); // ensure loading screen does not disappear too soon

  const [{ loading }, submit] = useAsyncFn(
    async (e: any) => {
      e.preventDefault();
      setIsLoading(true);
      const password = e.target.password.value;
      try {
        const res = await API.login({
          authPassword: { email, password },
        });
        if (!res.token) throw new Error();
        Event("AUTH", "password login", "success");
        logUser(res.token);
      } catch (e: any) {
        setIsLoading(false);
        const error = handleError(e.response?.data?.code, e.response?.data?.data?.email || "");
        if (error) setError(error);
      }
    },
    [router, email, logUser],
  );

  if (!email) return null;
  return (
    <div className={cls(styles.container, styles.half)}>
      <SEO title="Ravis de vous revoir !" />
      {isLoading ? (
        <Loader text="Connexion en cours..." />
      ) : (
        <>
          <Button priority="tertiary" size="small" iconId="fr-icon-arrow-left-line" onClick={() => router.back()}>
            Retour
          </Button>
          <div className={styles.title}>
            <h1>Ravis de vous revoir&nbsp;!</h1>
            <Tag
              className={styles.tag}
              dismissible
              nativeButtonProps={{
                onClick: () => router.push(getPath("/auth", "fr")),
              }}
            >
              {email}
            </Tag>
          </div>

          <form onSubmit={submit}>
            <PasswordInput
              label="Mot de passe"
              messages={
                !!error
                  ? [
                      {
                        message: error,
                        severity: "error",
                      },
                    ]
                  : []
              }
              messagesHint=""
              nativeInputProps={{ name: "password", autoFocus: true }}
            />
            <div className={cls("mt-2", styles.mb)}>
              <FRLink href={`/auth/reinitialiser-mot-de-passe?email=${email}`}>Mot de passe oublié&nbsp;?</FRLink>
            </div>

            <Button
              iconId="fr-icon-arrow-right-line"
              iconPosition="right"
              className={styles.button}
              nativeButtonProps={{ type: "submit" }}
              disabled={loading}
            >
              {!has2FA ? "Me connecter" : "Me connecter avec le mot de passe"}
            </Button>
          </form>

          {!has2FA && (
            <>
              <div className={cls(styles.separator, "my-10")}>
                <span>ou</span>
              </div>

              <Button
                linkProps={{ href: getPath("/auth/code-connexion", "fr", `?email=${email}`) }}
                className={styles.button}
                priority="tertiary"
              >
                Me connecter avec un code reçu par mail
              </Button>
            </>
          )}
        </>
      )}
    </div>
  );
};

export const getStaticProps = defaultStaticProps;
export default AuthLogin;

// override default layout and options
AuthLogin.getLayout = (page: ReactElement) => <Layout>{page}</Layout>;
