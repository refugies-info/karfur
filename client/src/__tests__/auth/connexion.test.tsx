import "jest-styled-components";
import mockRouter from "next-router-mock";
import { initialMockStore } from "__fixtures__/reduxStore";
import { wrapWithProvidersAndRenderForTesting } from "../../../jest/lib/wrapWithProvidersAndRender";
import connexion from "../../pages/auth/connexion";
import { setupGoogleMock } from "../../__mocks__/react-google-autocomplete";

jest.mock("next/router", () => require("next-router-mock"));

describe("auth/connexion", () => {
  beforeEach(() => {
    jest.clearAllMocks();
    setupGoogleMock();
  });

  it("renders null if no email", () => {
    const component = wrapWithProvidersAndRenderForTesting({
      Component: connexion,
      reduxState: {
        ...initialMockStore,
      },
    });
    expect(component).toMatchSnapshot();
  });
  it("renders connexion if email", async () => {
    await mockRouter.push("/auth/connexion?email=test@example.com");
    const component = wrapWithProvidersAndRenderForTesting({
      Component: connexion,
      reduxState: {
        ...initialMockStore,
      },
    });
    expect(component).toMatchSnapshot();
  });
});
