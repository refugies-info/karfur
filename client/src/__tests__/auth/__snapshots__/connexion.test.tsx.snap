// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`auth/connexion renders connexion if email 1`] = `
<div
  className="container half"
>
  <button
    className="fr-btn fr-btn--tertiary fr-btn--sm fr-icon-arrow-left-line fr-btn--icon-left"
    id="fr-button-:r6:"
    onClick={[Function]}
    style={{}}
  >
    Retour
  </button>
  <div
    className="title"
  >
    <h1>
      Ravis de vous revoir !
    </h1>
    <button
      className="fr-tag fr-tag--dismiss tag"
      id="fr-tag-:r7:"
      onClick={[Function]}
      style={{}}
    >
      test@example.com
    </button>
  </div>
  <form
    onSubmit={[Function]}
  >
    <div
      className="fr-password"
      id="password-input-:r8:"
    >
      <label
        className="fr-label"
        htmlFor="password-:r9:"
      >
        Mot de passe
      </label>
      <div
        className="fr-input-wrap"
      >
        <input
          autoFocus={true}
          className="fr-password__input fr-input"
          disabled={false}
          id="password-:r9:"
          name="password"
          type="password"
        />
      </div>
      <div
        className="fr-password__checkbox fr-checkbox-group fr-checkbox-group--sm"
      >
        <input
          aria-label="Afficher le mot de passe"
          id="password-:r9:-toggle-show"
          type="checkbox"
        />
        <label
          className="fr-password__checkbox fr-label"
          htmlFor="password-:r9:-toggle-show"
        >
          Afficher
        </label>
      </div>
    </div>
    <div
      className="mt-2 mb"
    >
      <a
        className="fr-link link"
        href="/auth/reinitialiser-mot-de-passe?email=test@example.com"
        onClick={[Function]}
        onMouseEnter={[Function]}
        onTouchStart={[Function]}
      >
        Mot de passe oublié ?
      </a>
    </div>
    <button
      className="fr-btn fr-icon-arrow-right-line fr-btn--icon-right button"
      disabled={false}
      id="fr-button-:ra:"
      style={{}}
      type="submit"
    >
      Me connecter
    </button>
  </form>
  <div
    className="separator my-10"
  >
    <span>
      ou
    </span>
  </div>
  <a
    className="fr-btn fr-btn--tertiary button"
    href="/auth/code-connexion?email=test@example.com"
    id="fr-button-:rb:"
    style={{}}
  >
    Me connecter avec un code reçu par mail
  </a>
</div>
`;

exports[`auth/connexion renders null if no email 1`] = `null`;
